browser/ui/views/ai_rewriter/ai_rewriter_button_view.cc

```diff
@@ -0,0 +1,201 @@
+                    views::InkDropHost::InkDropMode::ON);
+                CreateToolbarInkdropCallbacks(button, kColorToolbarInkDropHover,
+                                              kColorToolbarInkDropRipple);
+                views::InstallRoundRectHighlightPathGenerator(
+                    button, gfx::Insets(), kButtonRadius);
+              })))
+      .BuildChildren();
+}
+
+AIRewriterButtonView::~AIRewriterButtonView() = default;
+
+base::WeakPtr<AIRewriterButton> AIRewriterButtonView::MaybeCreateButton(
+    content::WebContents* contents) {
+  auto* browser = chrome::FindBrowserWithTab(contents);
+
+  // Possible for non-tab WebContents
+  if (!browser) {
+    return nullptr;
+  }
+
+  auto* button = new AIRewriterButtonView(browser, contents);
+
+  auto* parent_widget = views::Widget::GetWidgetForNativeWindow(
+      browser->window()->GetNativeWindow());
+  CHECK(parent_widget);
+
+  views::Widget::InitParams params(
+      views::Widget::InitParams::Type::TYPE_CONTROL);
+  params.parent = parent_widget->GetNativeView();
+  params.activatable = views::Widget::InitParams::Activatable::kNo;
+  params.delegate = button;
+  params.shadow_type = views::Widget::InitParams::ShadowType::kDrop;
+  params.corner_radius = kButtonRadius;
+  params.autosize = true;
+
+  auto* widget = new views::Widget();
+  widget->Init(std::move(params));
+  widget->Hide();
+
+  return button->weak_ptr_factory_.GetWeakPtr();
+}
+
+void AIRewriterButtonView::Show(const gfx::Rect& rect) {
+  CHECK(GetWidget());
+  GetWidget()->Show();
+
+  auto* browser_view = BrowserView::GetBrowserViewForBrowser(
+      chrome::FindBrowserWithTab(web_contents()));
+  CHECK(browser_view);
+
+  auto offset = browser_view->contents_container()->bounds().OffsetFromOrigin();
+
+  constexpr int kPaddingY = -8;
+  auto size = GetPreferredSize();
+  auto pos = rect.origin() + offset;
+  pos.Offset(GetPreferredSize().width() / 2,
+             -GetPreferredSize().height() / 2 + kPaddingY);
+  GetWidget()->SetBounds(gfx::Rect(pos, size));
+}
+
+void AIRewriterButtonView::Hide() {
+  GetWidget()->Hide();
+}
+
+void AIRewriterButtonView::Close() {
+  GetWidget()->Close();
+}
+
+bool AIRewriterButtonView::IsShowing() const {
+  return GetWidget()->IsVisible();
+}
+
+ui::WebDialogDelegate* AIRewriterButtonView::OpenDialog() {
+  auto* host = web_contents()->GetFocusedFrame()->GetRenderWidgetHost();
+  CHECK(host);
+
+  auto* host_view = host->GetView();
+  if (!host_view) {
+    return nullptr;
+  }
+
+  auto selected = host_view->GetSelectedText();
+  return AIRewriterDialogDelegate::Show(web_contents(),
+                                        base::UTF16ToUTF8(selected));
+}
+
+void AIRewriterButtonView::PrimaryPageChanged(content::Page& page) {
+  Close();
+}
+
+void AIRewriterButtonView::WebContentsDestroyed() {
+  Close();
+}
+
+void AIRewriterButtonView::OnVisibilityChanged(content::Visibility visibility) {
+  if (visibility == content::Visibility::HIDDEN) {
+    Hide();
```

Comment:
how it will work if you have two windows opened and just switch focus from one window to another?

does it work the way you expect it to?

Reply:
Seems to, yeah :smile:

[Screencast from 2024-08-27 09-00-05.webm](https://github.com/user-attachments/assets/a1c2a27a-a3ad-45bf-9161-19c7d14d5e7c)